sources/engine/Stride.Rendering/Rendering/Mesh.cs

```diff
@@ -76,5 +82,165 @@ public Mesh(MeshDraw meshDraw, ParameterCollection parameters)
+        {
+            var shape = Shapes?.Keys?.Where(c => c.Name == shapeName).FirstOrDefault();
+            if (shape != null)
+            {
+                SetBendShapeWeight(shape, weight);
+            }
+        }
+
+
+        public void SetBlendShapeWeightByIndex(int index, float weight)
+        {
+            var shape = Shapes?.ElementAt(index);
+            if (shape != null)
+            {
+                SetBendShapeWeight(shape.Value.Key, weight);
+            } 
+        }
+
+        void SetBendShapeWeight(Shape shape, float weight)
+        {
+            if (shape == null) { return; }
+            bool containsShape = Shapes != null && Shapes.ContainsKey(shape);
+            if (containsShape)
+            {
+                Shapes[shape] = weight;
+
+                if (MATBSHAPE != null)
+                {
+                    int index=Shapes.IndexOf(c => c.Key == shape);
+                    for (int i = 0; i < Draw.VertexMapping.Length; i++)
+                    {
+                        var shiftedIndex = index * Draw.VertexMapping.Length + i;
+                        MATBSHAPE[shiftedIndex][3] = weight;
+                    }
+                }
+            }   
+        }
+
+        public bool BlendShapeProcessingNecessary = false;
+
+       public Matrix[] MATBSHAPE { get; set; }
+
+       public float BasisKeyWeight { get; set; }
+
+       public void ProcessBlendShapes()
+       {
+            if (BlendShapeProcessingNecessary)
+            {
+                if (Shapes == null || Shapes.Count < 1) { return; }
+                var blendShapeVertices = AdjustPositionToDrawInstance();
+                var blendShapesWeights = GetBlendWeights(out float cummulativeWeight); ;
+                BasisKeyWeight = 1 - cummulativeWeight;
+
+                MATBSHAPE = new Matrix[Draw.VertexMapping.Length*blendShapesWeights.Length];
+                for (int iBendShape = 0; iBendShape < blendShapesWeights.Length; iBendShape++)
+                {
+                    for (int iVert = 0; iVert < Draw.VertexMapping.Length; iVert++)
+                    {
+                        var vectorValue = blendShapeVertices.ElementAt(iBendShape * Draw.VertexMapping.Length + iVert);
+                        MATBSHAPE[iBendShape * Draw.VertexMapping.Length + iVert].Row1 = new Vector4(vectorValue, blendShapesWeights[iBendShape]);
+
+                    }
+                }
+                BlendShapeProcessingNecessary = false;
+            }
+        }
+        
+
+        public void AddBlendShapes(Shape shape, float weight)
+        {
+            (Shapes ??= new()).Add(shape, weight); 
+            Parameters.Set(MaterialKeys.HasBlendShape, true);
+            
+        }
+
+        public float[] GetBlendWeights(out float cummulativeWeight)
+        {
+            float[] shapeWeights = new float[Shapes.Count];
+            cummulativeWeight = 0f;
+             
```

Comment:
Looks like this method is called fairly frequently, you could use a stackalloc'ed span in the caller of this method to avoid hammering the GC. Something like 
```cs
Span<float> blendShapesWeights = stackalloc float[Shapes.Count];
GetBlendWeights(blendShapesWeights, out float cummulativeWeight);


public void GetBlendWeights(Span<float> shapeWeights, out float cummulativeWeight)
{
   cummulativeWeight = 0f;
   // ...
```